src/main/java/org/elasticsearch/index/query/QueryBuilderSerializer.java

```diff
@@ -0,0 +1,62 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.index.query;
+
+import org.elasticsearch.common.inject.Inject;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.indices.query.IndicesQueriesRegistry;
+
+import java.io.IOException;
+
+/**
+ * Reads and writes a {@link QueryBuilder} when serialized over the wire based on its name which gets serialized
+ * before the the query itself and allows, when reading, to create the proper {@link QueryBuilder} subclass instance.
+ */
+public class QueryBuilderSerializer {
+
+    @Inject
+    static IndicesQueriesRegistry indicesQueriesRegistry;
```

Comment:
I think I personally don't like any of these solutions...they all feel like a hack to me. The staticness might not make much sense given that this should probably be per node, and multiple nodes can run in the same jvm, potentially even be part of different clusters? We have yet to find a better solution.